### ISSUE TITLE
The "Recommended releases" block do not appears on the pages of some projects.

### ISSUE THREAD

**Issue by findlabnet** [backdrop-ops/backdropcms.org#799]
So there is no regular download link on these pages.

For example:
- https://backdropcms.org/project/ban_ip
- https://backdropcms.org/project/abuseipdb_report
- https://backdropcms.org/project/antiscan
- etc.

**Comment by klonos** [comment]
Thanks for taking the time to report this @findlabnet 👍🏼 

I had a quick look, and this seems to be an attached view rather than a block. I could not tell what would cause this to happen though. Perhaps @jenlampton has a better idea.

**Comment by indigoxela** [comment]
Same for themes. It seems to be related to the date of latest release.

At first sight the link is missing in *all* projects with releases older than some months – the majority of projects, actually. I might have to do something with a views (contextual) filter based on dates?

**Comment by findlabnet** [comment]
>projects with releases older than some months

but this - https://backdropcms.org/project/seo_meta - almost 3 years old release - have this block.

**Comment by indigoxela** [comment]
Hm, then my suspicion was wrong. No clue, what the problem is then.

**Comment by findlabnet** [comment]
Someone built it, so someone know how it works.

**Comment by bugfolder** [comment]
That is a View-provided block, but it is not added to the page as a block; it is rendered via `hook_node_view()` for the project release content type.

Specifically, the View is "Project release download table" (`project_release_download_table`), which has 4 displays. The first, "Block" (`block`) provides this block. (The display "Other" (`other`) is an attachment to `block`.)

Rather than appearing on the page as a block in the usual way, though, it is added in `project_release_node_view()`, which calls `project_release_donwload_table()`, which calls `views_embed_view('project_release_download_table', 'block', $pid)` to produce the embedded View.

The filters on the `block` display are Published and whether the project release is the recommended branch. So it might be that there's a mismatch there.

If one of those with the Powers could update the sanitized dbs/files and update live-active/staging config, I will pull a fresh local copy and dig into this further.

**Comment by jenlampton** [comment]
> The filters on the block display are Published and whether the project release is the recommended branch. So it might be that there's a mismatch there.

I suspected the recommended branch could be part of the problem, but I checked on all the examples in the Original Post, and they all seem to have one. Perhaps it's a relationship issue?

**Comment by bugfolder** [comment]
Well that's pretty weird. I pulled a fresh copy of santized db and files, copied live-active into dev-active, so I (theoretically) have a local copy of the current state of b.org; and those problematic pages are showing the Recommended releases blocks just like they're supposed to.

**Comment by jenlampton** [comment]
Did you use `drush-uli` to log in? Maybe it's a permissions issue?

**Comment by bugfolder** [comment]
I didn't use drush-uli to log in, but made myself a Developer and logged in the usual way. But, thinking it might be permissions, I also checked the page in another browser as anonymous user, and it showed up there, too.

Oh yeah, and made sure codebase is freshly checked out.

**Comment by indigoxela** [comment]
> I also checked the page in another browser as anonymous user, and it showed up there, too.

That's odd. Did anyone try to flush all caches on b-org yet?

**Comment by bugfolder** [comment]
Also odd, possibly relevant: the repository shows that the last commit was 6 days ago; and after git checkout, live-active and staging still differ in several files.

**Comment by bugfolder** [comment]
This issue looks like https://github.com/backdrop-ops/backdropcms.org/issues/755 from January 29.

**Comment by indigoxela** [comment]
It seems, the solution earlier this year was to save the view once - without changes. Smells like a caching problem. So my question again: did anyone try to flush all caches?

The problems seems to pop up from time to time. And it's annoying for all affected contrib projects. Isn't there a more robust way to implement a download link?

**Comment by bugfolder** [comment]
Perhaps one of those with Developer role (e.g., @jenlampton, @quicksketch, @stpaultim, @herbdool, @jackaponte) could try the cache-flushing.

It's a little puzzling. The View itself doesn't have any caching turned on. Seems like it should be a perfectly robust way of implementing the download tables. A little frustrating that the problem doesn't manifest on local dev copy of the site.

**Comment by indigoxela** [comment]
> The View itself doesn't have any caching turned on.

Maybe no data cache, but the views options always get cached (in db table cache_views). If there are performance concerns regarding flushing the caches, go to /admin/structure/views/settings/advanced and use the button there – flushing views cache isn't available as individual option via the admin menu. Saving a single view (any) also rebuilds that whole cache.

If that fixes the problem, then ... we have to head scratch a little more, to figure out the reason why not all projects are affected.

**Comment by jenlampton** [comment]
I have flushed the views cache, but I don't have time to test today.

**Comment by bugfolder** [comment]
Thx @jenlampton. Alas, that didn't fix it.

**Comment by bugfolder** [comment]
And now that I can, I've flushed all caches, and that seems to have fixed it, at least, for the modules listed above (and a few others I'd spot-checked).

So, on to head-scratching, as to why that's necessary. (And kicking myself that I didn't think to try flushing the individual caches one at a time to narrow the list of suspects.)

**Comment by indigoxela** [comment]
Hm... there's still something fishy, I didn't check many projects, but quickly found two that don't have the block:

- https://backdropcms.org/project/formblock
- https://backdropcms.org/project/void_menu

And something else catches my eye: The pager below the modules list is missing on some pages.
Or does anyone else get the pager there?

- https://backdropcms.org/modules?s=&page=23
- https://backdropcms.org/modules?s=&page=20
- https://backdropcms.org/modules?s=&page=9
- https://backdropcms.org/modules?s=&page=4

Some pages do have a pager... Another odd thing without recognizable logic.

**Comment by bugfolder** [comment]
Even weirder. When I visit https://backdropcms.org/modules?s=&page=4 while logged in, I do see the pager, but its numbering is messed up. I'm on page 4, but the 5 is highlighted, and if, for example, I hover over 3, the tooltip shows me that the link actually goes to 2.

![pager](https://user-images.githubusercontent.com/29214084/122976866-b3599d00-d349-11eb-8c10-d83952f6115e.png)

But if I visit that page in an incognito window (so not logged in), there's no pager at all (as you observe).

**Comment by findlabnet** [comment]
Some time ago this part of the site was redesigned / rebuilded. The changes also affected the general behavior: earlier - any new (latest) release is displayed first in the list, now the items are sorted by the date of the first (not current) release.

**Comment by indigoxela** [comment]
Another finding re pager...

I manually changed the url from `https://backdropcms.org/modules?s=&page=9` to `https://backdropcms.org/modules?page=9` and - tadaa the pager is there. So, the `s=` breaks the pager (the search).

Verified with `https://backdropcms.org/modules?s=menu` (first page, but with search term) - pager gone. Should I open an new issue for that?

**Comment by jenlampton** [comment]
Yes to a new issue, and I recognize that extra s - we should check the
custom google analytics tracking code to see if it's to blame. It is added
specifically on the project search pages to track what people are searching
for, so would explain the strange behavior there.

I am convinced this pager stuff is a different issue than the releases
visibility.

**Comment by indigoxela** [comment]
> Yes to a new issue

OK, this is driving me nuts... :confused: Today it all works again. I just repeated the exact same steps (for a new report). Pager works fine today. I'd like to note that I'm always anonymous - I don't have an account on backdropcms-org.
Did anyone sprinkle some fairy dust over the webserver?

**Comment by indigoxela** [comment]
Time for some conspiracy theory: could it be some off-by-one thing? I checked a lot more modules, and it seems to me that the affected ones all have just *one* stable release. I could imagine some `$cnt > 1` instead of `$cnt >= 1`...

- https://backdropcms.org/project/formblock
- https://backdropcms.org/project/void_menu
- https://backdropcms.org/project/dblog_pagesplit
- https://backdropcms.org/project/taxonomy_set_lineage
- https://backdropcms.org/project/google_tag
- https://backdropcms.org/project/menu_link_weight
- https://backdropcms.org/project/ckeditor_accordion
- https://backdropcms.org/project/schema

I also found this one: https://backdropcms.org/project/me at first sight it seems to have three releases, but all of them are pre-releases: https://github.com/backdrop-contrib/me/releases

So my wild thesis still needs refutation. :grin:

**Comment by indigoxela** [comment]
Forget my previous comment, I already found refutation - https://backdropcms.org/project/multiple_email

Back to caching theory - which also doesn't explain, why some modules are affected, others aren't and after flushing caches it seems to be different ones.

**Comment by jenlampton** [comment]
I wonder if its a cache bin issue on the server? Are we using mwmcache?

**Comment by bugfolder** [comment]
A little bit of narrowing down to report. The Form block module page was not showing the recommended releases on the project page, https://backdropcms.org/project/formblock,

but was showing the release in search results, https://backdropcms.org/modules?s=Form+block.

Checking the View that produces that block, (https://backdropcms.org/admin/structure/views/view/project_release_download_table, display "Block", use contextual filter = 110 for preview) shows it gives the correct result in preview. So the View configuration looks OK.

Then using the admin bar command _Admin > Home > Flush all caches > Page and else_ restored its appearance on the project page.

Now, it appears that when a new release is created on Github, the sequence of calls includes a flush of the cache of the project release download table, but I didn't see anywhere that it flushed the cache of the page on which that table appears. So that might be part of the puzzle.

**Comment by bugfolder** [comment]
OK, here's what appears to be going on. As noted above, the `project_release_download_table` that appears on the project page is being cached. There are various pathways that _can_ result in that cache getting cleared, but in following the pathway that starts with `_project_github_create_project_release()`, I'm not seeing anything that would clear that particular cache. (I can't actually debug-step through this locally because that call depends on a new release being pushed from GitHub.)

The logical place for that caching to get initiated within `_project_github_create_project_release()` would be in the code
```
    // If the release is saved successfully, update the project node to trigger
    // a rebuild of the current version information.
    if ($success) {
      $project_node->updated = REQUEST_TIME;
      $project_node->save();
    }
```
since saving `$project-node` would (for example) rebuild any cached version of the node or its page. But the `project_release_download_table` that appears on the project page is cached separately. So it should (probably) be cleared separately as well.

I think, then, that a solution would be to explicitly clear this cached version in `project_github_node_presave()`, and so will submit a PR with that change.

Only way I can think of to test its effectiveness is to actually deploy it live, wait a little while for some new releases, then check to see if the project pages are getting updated (and existing project pages aren't haywire). If it turns out not to be the fix, it's easily reverted, since the change is small.

**Comment by indigoxela** [comment]
@bugfolder thank you so much for digging!

Did you flush all caches again? Or is this problem bound to releases and the block won't appear until a new module release is out?
I'm asking because there are still modules that do have a stable release, where the block doesn't appear.

**Comment by bugfolder** [comment]
Still digging, but it looks like I haven't hit bottom yet, alas.

My theory was that that block was getting generated initially empty, then that was being cached and never updated. So my PR would update the block (by clearing its cache) each time a new release was generated. But it wouldn't go back and fix any of the existing problematic pages.

And doing a little more experimentation on the existing project pages seems to indicate that there's more going on with the existing pages that are missing the block. I did the following with the Schema project page, which was one of the ones you noted had a missing block. Viewing both in a logged-in window (as a Developer) and a private (non-logged-in) window.

View the page: block is missing in both windows.

Flush "page and else" cache: same thing, block still missing in both windows. (This tells me that there's more going on than the caching of the table, because I'd previously verified that the "and else" of "page and else" clears the `cache_entire project_release_download_table` table.)

Click the "edit" tab of the project node and save immediately: The block now correctly appears in the logged in window. But it doesn't appear in the anonymous window after refreshing. However, after a few minutes, another refresh, and it does appear. (This is probably due to the standard caching of anonymous pages with a 5-minute lifetime.)

So edit-and-save appears to update a problematic node page. But what extra that does that ordinary cache-flushing doesn't is still unclear.

**Comment by indigoxela** [comment]
So we have at least two things going on. The page cache for anonymous seems less problematic to me, waiting some minutes until the block appears is OK.

That it's necessary to edit/save the node to get the block even for logged-in - after the whole table got flushed - seems odd. Entity cache? (Wild guessing... I'm not familiar with the project module.)

**Comment by findlabnet** [comment]
Can someone step in?

**Comment by bugfolder** [comment]
Came back to this.

This was one ofthe pages @indigoxela reported was missing its Recommended release block in https://github.com/backdrop-ops/backdropcms.org/issues/799#issuecomment-866887663 above.

https://backdropcms.org/project/formblock

I visited that page as a logged-in user, and the Recommended releases block was indeed missing. Cleared all caches on b.org, and it appeared.

Built my local copy of b.org from the latest sanitized dbs and the Recommended release block was present from the outset (even before any explicit cache-clearing), both as anonymous visitor and authenticated.

So pretty Heisenbuggy. Also hard to diagnose when it doesn't manifest on local versions of the site.

**Comment by findlabnet** [comment]
I can confirm that after you've cleared the caches on b.org, blocks appear on pages that weren't there a few hours ago, both for anonymous visitor and authenticated.
It is very sad to see this error on the pages presenting our community to the world.

**Comment by bugfolder** [comment]
Yes. I'd like to see if my PR fixes the problem on the live site. Perhaps you could review the PR to help it on its way to adoption?

**Comment by findlabnet** [comment]
Unfortunately, I have no way to test this PR, but at least it does not seem too destructive.

**Comment by jenlampton** [comment]
I agree that the risk is low, so I've merged the PR and pushed up to b.org. 🤞

**Comment by bugfolder** [comment]
Reopening because the problem persists.

A temporary fix is to clear all caches.

I built a little tool that lets me scrape several project pages at any time, looking for the "Recommended releases" block. It verified that for several projects (Ubercart, References, Rules Once per Day, Elysia Cron),

- they were missing releases yesterday;
- Clearing all caches restored the releases;
- Today they were missing again for the latter 3;
- Clearing all caches restored the releases again.

**Comment by indigoxela** [comment]
Odd. What happens in background (which jobs), that make them vanish randomly? Random bugs are the worst ones.

**Comment by bugfolder** [comment]
Yup. 

In my diagnostic tool, I expanded the list of module pages to scrape to 30. Overnight, 25 of them changed to show no releases. Once again, clearing all caches put them back.

So the fundamental issue is that at some point, a call to `project_release_download_table()` is getting made that results in an empty view being generated and cached, and then that content gets served up until the next cache-clear. 

Things that could conceivably trigger such a sporadic problem include:

* Updates due to GitHub calls (nodes get updated every time there's a change from Github). This seems unlikely, though, to affect large numbers of modules, including ones that I know haven't seen any Github activity.
* Cron calls that expire caches and/or do other things. Seems like a pretty broad category.
* User visits to particular pages that somehow trigger generation of the empty view.

I did observe that any project search that includes a module in its results has the side effect of creating (or retrieving from cache) the download release table for that module, even though that table isn't displayed in the search results. I think this generation can be safely removed (thereby slightly speeding up project searches), and that will reduce the possible code paths that generate/cache this table. So I'll put in a PR for that, and also revert [PR 800](https://github.com/backdrop-ops/backdropcms.org/pull/800) because it didn't have any discernible effect on the problem.

**Comment by bugfolder** [comment]
So I just noticed that this issue is broader than just the vanishing "Recommended releases" block on the project page. On my local dev site, I checked the "available updates" page and it shows "no releases available" for many contrib modules. Cleared all caches on b.org, re-ran "check available updates" on the local site, and they all showed up.

So the problem is deeper than the mere caching of that particular view display. Need to go look into what, exactly, "check available updates" queries. That's another avenue for trying to pin this down.

**Comment by findlabnet** [comment]
FTR. Missing blocks. Again.

**Comment by quicksketch** [comment]
Thanks @findlabnet. I deployed an update to Project module yesterday but I didn't think it would affect the blocks. Here's the PR: https://github.com/backdrop-ops/backdropcms.org/pull/819/files

There is some debugging code removed as part of the update, but the output seems as though it would have been the same.

**Comment by findlabnet** [comment]
Caches was not flushed at this (update) time? Usually it help.

**Comment by bugfolder** [comment]
I flushed the cached right after the previous exchange, which restored all of the missing releases (as expected).

Although the Project mode update @quicksketch mentions did indeed wipe out the debugging code (which I'd been checking periodically), since putting in the debugging code, I hadn't seen any releases vanish, so the debugging code didn't provide any new insights.

I run a test page each day that scrapes 30 project pages on b.org, looking for the releases (or their lack). Prior to putting in the debug code, several vanished each day (to be restored by a manual cache-clear.) Once the debug code was in place, they quit vanishing. Post-[PR 819](https://github.com/backdrop-ops/backdropcms.org/pull/819/files), they still haven't vanished (it's now been 7 days).

Whatever the issue was, it seems to be in abeyance, at least, for a while.

I blame Zen.CI. ;o)

**Comment by bugfolder** [comment]
So, after 12 straight days of no vanishing releases, checked this morning, and 24 of 31 releases vanished.

No obvious reasons visible in watchdog, though there is an oddity: most (but not all) of the hourly cron runs are failing with "Cron could not run because an invalid key was used." This has been happening for a few days, though.

As usual, clearing the "Page and else" caches restores the releases to the individual project pages.

**Comment by findlabnet** [comment]
Just confirming.

**Comment by bugfolder** [comment]
...And we're back to "each day a whole slew of modules lose their releases" (today, 10 of 31 missing), curable (temporarily) by cache-clearing.

So, at some point the cache gets cleared, and then the next time that `project_release_download_table()` gets called for that project, the View is empty for some mysterious reason, and then the empty View hangs around forever in the cache (or until the next cache-clearing).

And the issue is indeed that the View is empty; the `div` wrappers of the View are present in the page HTML (and in the cache); they just don't contain anything.

A hacky workaround is to not cache the View output unless it contains at least some release. That wouldn't solve the underlying problem, but at least it would mean that users will see releases most of the time. (And for the handful of projects that truly have no releases, there'd be no caching, but then, they're probably not visited very often.)

**Comment by herbdool** [comment]
Maybe it's ok to disable Views caching for this View until we know more. At least there will still be page caching and most visitors will be anonymous.

**Comment by bugfolder** [comment]
Caching is already turned off in each of the displays of [this View](https://backdropcms.org/admin/structure/views/view/project_release_download_table/configure/block_1). The caching I was referring to is a separate cache within function `project_release_download_table ()`.

This morning's check revealed 0/31 vanishing releases (yay).

**Comment by bugfolder** [comment]
So, 4 days with no vanishes, then today, 17 of 31 releases vanished.

Gonna try the "hacky workaround."

**Comment by indigoxela** [comment]
@bugfolder one day, when you solved this Heisenbug, you'll get the Nobel Prize.

Is there anything from "outside", that influences the vanishing Problem. For instance, when there are new releases (contrib), the table or the update job chokes on something...

(Sure, wild guessing without any proper ideas...)

**Comment by bugfolder** [comment]
Yeah. I've looked through the watchdog log for anything suspicious to no avail. My previous round of diagnostic code got wiped out with the last update of project module, so there's not much to go on just now, but I'll update the diagnostics code shortly.

**Comment by jenlampton** [comment]
@bugfolder once you have your diagnostic code ready, can you create a PR against the repo (with only those changes?) We can document this in our PATCHES.txt file so that the patch can be re-applied after a module update.

Most people doing any kind of module update will check that PATCHES.txt file to make none of the changes made to our site get lost.

**Comment by jenlampton** [comment]
@bugfolder patch applied, code deployed, and PATCHES.md updated :)

**Comment by bugfolder** [comment]
Hey, we have another small data point: I just ran the project-scrape, and 11 of 31 releases had vanished. That's since this morning (7 hours ago).

And it (presumably) happened BEFORE you deployed the new code, because there's no report of the vanishing in watchdog (as the diagnostic code would do).

As usual, cleared the cache to restore things. Now we wait for further vanishment.

**Comment by bugfolder** [comment]
> one day, when you solved this Heisenbug, you'll get the Nobel Prize.

I'm shopping for tickets to Stockholm.

TL;DR: root cause was conflicts between pagers in Views that display other Views. PRs will follow shortly.

For the full story, settle into a comfortable chair and read on.

First, and what cracked this open: the diagnostics identified a "toxic" URL that reliably reproduced the problem (hence enabling local debugging). To see it in action:

* Clear caches.
* Visit https://backdropcms.org/modules/?s=&page=1
* Visit https://backdropcms.org/admin/reports/dblog to see the result.

The current diagnostics throw 25 reports of missing releases into watchdog. (Current diagnostics also disable caching of blank results, so you won't have poisoned the cache.)

Note that you need to include the "&page=1" in the URL to cause the problem. For example, this is harmless:

* https://backdropcms.org/modules/?s=

It doesn't have to be page 1 (which is actually the second page of results), and you could include actual search terms—for example, this also causes the problem:

* https://backdropcms.org/modules/?s=module&page=2

But you need to pick a search term that generates enough results that the specified results page exists in order to cause the problem.

Prior to the latest cache-disabling/diagonostics, the releases would go missing for a project if the cache has been recently cleared and the project release download table (PRDT) is generated for the project on the module search results page. However, if the cache was cleared and then you view the individual project page, the PRDT is generated properly. The reason it is sometimes missing from the individual project page is that the empty PRDT generated by a toxic search URL was getting cached and then rendered on the individual project page.

So this is part of what made this so mysterious: the empty PRDT was getting generated on the _module search pages_ (where it's not actually displayed), not on the individual project page where it's actually displayed. This is also why errors came in batches, since every project on the search results page ended up with an empty cached PRDT.

And this toxic URL was only generated by web crawlers and couldn't be generated by user visitation. How do we know this? Because _the pager is currently missing at the bottom of the search results page_. So there's no way a user could visit the toxic URL—there's no pager to click on. (Now. Evidently it was possible sometime in the past, which is how the address got into web crawlers' records.)

Now, a puzzling thing about this is, although the PRDT is getting generated for each project on a search results page, it isn't actually shown on the search results page. Debugging locally revealed that it was being called when each project was being rendered via `project_release_node_view()` using the `'project_search'` view mode. But in this view mode, it doesn't need to be generated. So a simple code change to not generate it in this view mode will fix this (and as a side effect, speed up the page rendering, since we're not rendering two more views for every project on the page).

But that leaves a couple open questions:

* First, why are the PRDT Views rendering blank when they're called from a search results page (as part of the rendering of the search results View), but they render properly when they're called to render the individual project pages?

* Second, what happened to the pager on these search results pages? The search results Views (machine name `modules`) specify pagers with 25 elements, but even if you use a very generic search term like "module" or "image" that should have more than 25 results, no pager appears.

Now, one of the settings for Views pagers is the "Pager ID", whose help text says:

> Unless you're experiencing problems with pagers related to this view, you should leave this at 0. If using multiple pagers on one page you may need to set this number to a higher value so as not to conflict within the ?page= array. Large values will add a lot of commas to your URLs, so avoid if possible.

Well, by golly we're experiencing problems with pagers. And indeed, if we change the Pager ID on the search results View from 0 to 1, the pager magically reappears (when appropriate) on the search results page.

So, obviously, the search results pager is conflicting with some other pager on the same page that also has ID 0. But where?

Well, remember that we're generating PRDTs for each project on the page. So we go check the settings of the various displays of the PRDT, and what do you know, the displays that are being rendered have pagers.

But they don't need them! They only ever show a small number of entries. These are the displays:

* Recommended (main) - shows the recommended release on the project page. Currently has a pager, but this will only ever be 1 item.
* Other (main) — shows the releases with the highest major version number that's not the recommended release. Currently has a pager, but will only be 1 or a few at most entries.
* Recommended (search) — this shows the recommended release on the search results page. It doesn't have a pager, so it's not affecting anything.
* Versions (sidebar) — this creates the "See all versions" link in the sidebar on the project page. It currently has a pager, but it will only ever contain a single item, so it doesn't need one.

And that make it clear what's happening and why blank PRDTs were getting generated! Here's what happens:

* Visitor goes to https://backdropcms.org/modules/?s=&page=1, i.e., asking for the second page of results of this module search.
* The module search View does its search, creates the list of modules, and starts to render them via `project_release_node_view()`.
* The rendering calls the individual PRDT Views, but _it passes `page=1` to their pagers_ because all the pagers have ID 0.
* Since none of these Views have a second page of results, the resulting View is blank.
* This blank value gets cached and sticks around until caches are cleared.

So, the solution to the bug is simply to remove the pagers from the `project_release_download_table` View. No more superfluous pagers, no more pager ID conflicts, no more vanishing cached releases.

There is still the issue that `project_release_node_view()` was rendering the PRDT when it didn't need to. There's no need to render it with the `project_search` view mode (it is already not rendered for `rss` and `search_index` view modes), so we might as well remove that rendering and speed up the search results page.

As they say, "well, there it is." A PR for backdropcms.org will be following, and I'll also file a PR against the contrib `project` module with the code changes.

**Comment by jenlampton** [comment]
Wowsers! Thank you so much for your continued effort on this problem :)
